src/App.tsx

```diff
@@ -0,0 +1,795 @@
+  }, []);
+
+  const [collection, setCollection] = useState<string | undefined>();
+  const [collectionData, setCollectionData] = useState<urlUtils.CollectionData | undefined>();
+  const [dataset, setDataset] = useState<Dataset | null>(null);
+  const [datasetName, setDatasetName] = useState("");
+  const [featureName, setFeatureName] = useState("");
+  const [selectedTrack, setSelectedTrack] = useState<Track | null>(null);
+  const [currentFrame, setCurrentFrame] = useState<number>(0);
+
+  const [isInitialDatasetLoaded, setIsInitialDatasetLoaded] = useState(false);
+  const [datasetOpen, setDatasetOpen] = useState(false);
+
+  const [colorRamp, setColorRamp] = useState(colorRamps[DEFAULT_COLOR_RAMP]);
+  const [colorRampMin, setColorRampMin] = useState(0);
+  const [colorRampMax, setColorRampMax] = useState(0);
+  const [isColorRampRangeLocked, setIsColorRampRangeLocked] = useState(false);
+  const [hideValuesOutOfRange, setHideValuesOutOfRange] = useState(false);
+  const [showTrackPath, setShowTrackPath] = useState(false);
+
+  const timeControls = useMemo(() => {
+    return new TimeControls(canv);
+  }, []);
+  const recordingControls = useMemo(() => {
+    return new RecordingControls(canv);
+  }, []);
+
+  // Recording UI
+  const [imagePrefix, setImagePrefix] = useState("");
+  const [useDefaultPrefix, setUseDefaultPrefix] = useState(true);
+  const [startAtFirstFrame, setStartAtFirstFrame] = useState(false);
+  /** The frame selected by the time UI. Changes to frameInput are reflected in
+   * canvas after a short delay.
+   */
+  const [frameInput, setFrameInput] = useState(0);
+  const [findTrackInput, setFindTrackInput] = useState("");
+  const [hoveredId, setHoveredId] = useState<number | null>(null);
+
+  // UTILITY METHODS /////////////////////////////////////////////////////////////
+
+  /**
+   * Copy the current collection, dataset, feature, track, and frame information
+   * to the page URL.
+   */
+  const updateUrl = useCallback((): void => {
+    urlUtils.saveParamsToUrl(
+      collection || null,
+      datasetName,
+      featureName,
+      selectedTrack ? selectedTrack.trackId : null,
+      currentFrame
+    );
+  }, [collection, datasetName, featureName, selectedTrack, currentFrame]);
+
+  /**
+   * Handles updating the canvas from state (for some common properties) and initiating the canvas
+   * render. Also updates other UI elements, including the plot and URL.
+   */
+  const drawLoop = useCallback(async (): Promise<void> => {
+    // Note: Selected track, frame number, etc. are not updated here.
+    // Those operations are async, and need to complete before a state update to be
+    // rendered correctly.
+    canv.setShowTrackPath(showTrackPath);
+    canv.setHideValuesOutOfRange(hideValuesOutOfRange);
+    canv.setColorRamp(colorRamp);
+    canv.setColorMapRangeMin(colorRampMin);
+    canv.setColorMapRangeMax(colorRampMax);
+    canv.setSelectedTrack(selectedTrack);
```

Comment:
When canv is a react component all this will be gone

Reply:
🔥🔥🔥🔥🔥